Step2 연관 관계 매핑 코드리뷰 요청드립니다.

dongha-byun opened the pull request:
안녕하세요.

우아한 테크캠프 Pro 5기 교육생 변동하 입니다.

step2 연관관계 매핑 미션에 대한 코드리뷰를 요청드립니다.

감사합니다.

dongha-byun committed:
Feat: 11.3 수업 자료 커밋

dongha-byun committed:
Refactor: 연관관계 편의 메서드 추가
- 11. 3 수업 내용

dongha-byun committed:
Feat: Add Entity relationship
- User : Question (1 : N)
- Question : Answer (1 : N)
- User : Answer (1: N)
- User : DeleteHistory (1 : N)

dongha-byun committed:
Chore : change code style (space after extends BaseEntity)

dongha-byun committed:
Refactor: refactor repository parameter

dongha-byun committed:
Test : add and refactor test code for repository

joojimin commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -15,14 +17,18 @@
     private Long id;
 
-    private Long writerId;
+    @ManyToOne(fetch = FetchType.LAZY)
```
`LAZY` 지연로딩 적용 👍

joojimin commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -39,7 +39,7 @@ public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteE
         }
 
-        List<Answer> answers = answerRepository.findByQuestionIdAndDeletedFalse(questionId);
+        List<Answer> answers = answerRepository.findByQuestionAndDeletedFalse(question);
```
객체지향적인 설계를 연습하는만큼 연관관계를 통해서 리스트를 불러오면 어떨까요~?
추가로 일급컬렉션도 활용한다면 아래 작성자 검사도 내부로직으로 깔끔하게 설계할 수 있을 것 같아요!

dongha-byun replied:
디테일한 조언 감사합니다! 고민해보고 수정해보도록 하겠습니다...!

joojimin commented on src/test/java/qna/domain/DeleteHistoryRepositoryTest.java:
```diff
@@ -47,4 +53,29 @@ void findAllByContentTypeTest1(){
+        List<DeleteHistory> historiesByUser = user.getDeleteHistories();
+
+        //then
+        assertThat(histories.size()).isEqualTo(3);
```
`Assertion.assertThat`은 검사하려는 대상을 감싼 후 다양한 검사를 지원하는데요!
해당 컬렉션의 사이즈를 검사하는 `hasSize` 메서드를 사용한다면 함수이름 자체가 로직을 설명할 수 있어 가독성에 더 도움이 될거같아요!

dongha-byun replied:
넵!! Assertions 도 더 공부해봐야겠네요 ㅠㅠ 알려주셔서 감사합니다! 수정해볼게요!

joojimin reviewed (changes requested):
안녕하세요~ 동하님!
미션진행 고생하셨습니다!
몇가지 확인하고 넘어가면 좋을법한 코멘트 드렸는데요!
확인 후 리뷰 재요청 부탁드립니다~!

dongha-byun committed:
Test: Assertions 메서드 수정

dongha-byun committed:
Refactor: QnaService Refactor
- List<Answer> refactor (AnswersOfQuestion)
- refactor relation mapping about question entity

joojimin commented on src/main/java/qna/domain/AnswersOfQuestion.java:
```diff
@@ -0,0 +1,36 @@
+    private AnswersOfQuestion(){}
+
+    public static AnswersOfQuestion create(List<Answer> answers){
+        AnswersOfQuestion answersOfQuestion = new AnswersOfQuestion();
```
생성자를 통해 인자를 받는게 어떨까요~?
`List<Answer> answers;`에도 `private final`이 같이 붙을 수 있어 좋을거같아요!

joojimin reviewed (approved):
안녕하세요~ 동하님!
피드백 반영 고생하셨습니다!
코멘트 하나 남겼는데 다음 미션 진행하시면서 확인해보면 좋을거같아요!
화이팅입니다!
